staging/src/k8s.io/pod-security-admission/test/fixtures_hostPorts.go

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/pod-security-admission/api"
+)
+
+/*
+TODO: include field paths in reflect-based unit test
+
+containerFields: []string{
+	`securityContext.capabilities.add`,
+},
+*/
+
+// ensureHostPorts ensures the pod and all initContainers and containers have a non-nil Ports.
+func ensureHostPorts(p *corev1.Pod) *corev1.Pod {
+	for i := range p.Spec.Containers {
+		if len(p.Spec.Containers[i].Ports) == 0 {
+			p.Spec.Containers[i].Ports = []corev1.ContainerPort{}
+		}
+	}
+	for i := range p.Spec.InitContainers {
+		if len(p.Spec.InitContainers[i].Ports) == 0 {
+			p.Spec.InitContainers[i].Ports = []corev1.ContainerPort{}
+		}
+	}
+	return p
+}
+
+func init() {
+	fixtureData_1_0 := fixtureGenerator{
+		expectErrorSubstring: "forbidden host ports",
+		generatePass: func(p *corev1.Pod) []*corev1.Pod {
+			p = ensureHostPorts(p)
+			return []*corev1.Pod{
+				p,
+				// no host ports
+				tweak(p, func(p *corev1.Pod) {
+					p.Spec.Containers[0].Ports = []corev1.ContainerPort{
+						corev1.ContainerPort{
+							ContainerPort: 12345,
+						},
+					}
+				}),
+				tweak(p, func(p *corev1.Pod) {
+					p.Spec.InitContainers[0].Ports = []corev1.ContainerPort{
+						corev1.ContainerPort{
+							ContainerPort: 12346,
+						},
+					}
+				}),
```

Comment:
a single testcase that puts ports in the container and initContainer is probable sufficient and will cut down on excessive fixtures. we mostly care about individual addressability in the failure cases to make sure we're specifically failing on each scenario

Reply:
Good point! Fixed it.